// typegoose/lib/typegoose.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var mongoose = require("mongoose");
require("reflect-metadata");
var data_1 = require("./data");
__export(require("./method"));
__export(require("./prop"));
__export(require("./hooks"));
__export(require("./plugin"));
__export(require("."));
var utils_1 = require("./utils");
exports.getClassForDocument = utils_1.getClassForDocument;
var Typegoose = (function () {
    function Typegoose() {
    }
    Typegoose.prototype.getModelForClass = function (t, _a) {
        var _b = _a === void 0 ? {} : _a, existingMongoose = _b.existingMongoose, schemaOptions = _b.schemaOptions, existingConnection = _b.existingConnection;
        var name = this.constructor.name;
        if (!data_1.models[name]) {
            this.setModelForClass(t, {
                existingMongoose: existingMongoose,
                schemaOptions: schemaOptions,
                existingConnection: existingConnection,
            });
        }
        return data_1.models[name];
    };
    Typegoose.prototype.setModelForClass = function (t, _a) {
        var _b = _a === void 0 ? {} : _a, existingMongoose = _b.existingMongoose, schemaOptions = _b.schemaOptions, existingConnection = _b.existingConnection;
        var name = this.constructor.name;
        var sch = this.buildSchema(t, name, schemaOptions);
        var parentCtor = Object.getPrototypeOf(this.constructor.prototype).constructor;
        while (parentCtor && parentCtor.name !== 'Typegoose' && parentCtor.name !== 'Object') {
            sch = this.buildSchema(t, parentCtor.name, schemaOptions, sch);
            parentCtor = Object.getPrototypeOf(parentCtor.prototype).constructor;
        }
        var model = mongoose.model.bind(mongoose);
        if (existingConnection) {
            model = existingConnection.model.bind(existingConnection);
        }
        else if (existingMongoose) {
            model = existingMongoose.model.bind(existingMongoose);
        }
        data_1.models[name] = model(name, sch);
        data_1.constructors[name] = this.constructor;
        return data_1.models[name];
    };
    Typegoose.prototype.buildSchema = function (t, name, schemaOptions, sch) {
        var Schema = mongoose.Schema;
        if (!sch) {
            sch = schemaOptions ? new Schema(data_1.schema[name], schemaOptions) : new Schema(data_1.schema[name]);
        }
        else {
            sch.add(data_1.schema[name]);
        }
        var staticMethods = data_1.methods.staticMethods[name];
        if (staticMethods) {
            sch.statics = Object.assign(staticMethods, sch.statics || {});
        }
        else {
            sch.statics = sch.statics || {};
        }
        var instanceMethods = data_1.methods.instanceMethods[name];
        if (instanceMethods) {
            sch.methods = Object.assign(instanceMethods, sch.methods || {});
        }
        else {
            sch.methods = sch.methods || {};
        }
        if (data_1.hooks[name]) {
            var preHooks = data_1.hooks[name].pre;
            preHooks.forEach(function (preHookArgs) {
                var _a;
                (_a = sch).pre.apply(_a, preHookArgs);
            });
            var postHooks = data_1.hooks[name].post;
            postHooks.forEach(function (postHookArgs) {
                var _a;
                (_a = sch).post.apply(_a, postHookArgs);
            });
        }
        if (data_1.plugins[name]) {
            for (var _i = 0, _a = data_1.plugins[name]; _i < _a.length; _i++) {
                var plugin = _a[_i];
                sch.plugin(plugin.mongoosePlugin, plugin.options);
            }
        }
        var getterSetters = data_1.virtuals[name];
        if (getterSetters) {
            for (var _b = 0, _c = Object.keys(getterSetters); _b < _c.length; _b++) {
                var key = _c[_b];
                if (getterSetters[key].options && getterSetters[key].options.overwrite) {
                    sch.virtual(key, getterSetters[key].options);
                }
                else {
                    if (getterSetters[key].get) {
                        sch.virtual(key, getterSetters[key].options).get(getterSetters[key].get);
                    }
                    if (getterSetters[key].set) {
                        sch.virtual(key, getterSetters[key].options).set(getterSetters[key].set);
                    }
                }
            }
        }
        var indices = Reflect.getMetadata('typegoose:indices', t) || [];
        for (var _d = 0, indices_1 = indices; _d < indices_1.length; _d++) {
            var index = indices_1[_d];
            sch.index(index.fields, index.options);
        }
        return sch;
    };
    return Typegoose;
}());
exports.Typegoose = Typegoose;
