// typegoose/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var data_1 = require("./data");
exports.isPrimitive = function (Type) {
    return ['String', 'Number', 'Boolean', 'Date', 'Decimal128', 'ObjectID'].includes(Type.name);
};
exports.isObject = function (Type) {
    var prototype = Type.prototype;
    var name = Type.name;
    while (name) {
        if (name === 'Object') {
            return true;
        }
        prototype = Object.getPrototypeOf(prototype);
        name = prototype ? prototype.constructor.name : null;
    }
    return false;
};
exports.isNumber = function (Type) { return Type.name === 'Number'; };
exports.isString = function (Type) { return Type.name === 'String'; };
exports.initAsObject = function (name, key) {
    if (!data_1.schema[name]) {
        data_1.schema[name] = {};
    }
    if (!data_1.schema[name][key]) {
        data_1.schema[name][key] = {};
    }
};
exports.initAsArray = function (name, key) {
    if (!data_1.schema[name]) {
        data_1.schema[name] = {};
    }
    if (!data_1.schema[name][key]) {
        data_1.schema[name][key] = [{}];
    }
};
exports.getClassForDocument = function (document) {
    var modelName = document.constructor.modelName;
    return data_1.constructors[modelName];
};
